app/gui/src/project-view/stores/project/index.ts

```diff
@@ -327,6 +328,86 @@ export function createProjectStore(
     })
   }
 
+  // Maximum number of in-progress expressions.
+  const MAX_IN_PROGRESS = 5
+
+  const inProgress = ref(0)
+  const queueLength = ref(0)
+
+  function queuedExecuteExpression(
+    expressionId: ExternalId,
+    expression: string,
+    timeoutMs: number = 5000,
+  ): Promise<Result<any> | null> {
+    if (inProgress.value > MAX_IN_PROGRESS) {
+      if (timeoutMs < 0) {
+        return Promise.reject(Err(`queuedExecuteExpression: Execution timed out.`))
+      }
+
+      queueLength.value += 1
+      const pause = queueLength.value * 250
+      return new Promise((resolve) => setTimeout(resolve, pause)).then(() => {
+        queueLength.value -= 1
+        return queuedExecuteExpression(expressionId, expression, timeoutMs - pause)
```

Comment:
This timeout behaviour is complex--an execution will be cancelled if any particular pause time would be greater than its timeout, but it may be paused multiple times; if the queue is medium-length and busy, an execution may be paused for arbitrarily long (in total) before being started or cancelled. Behaviour would be more predictable if a deadline were computed when the request is first paused.